tritium-lib/src/main/java/com/palantir/tritium/proxy/ByteBuddyInstrumentation.java

```diff
@@ -0,0 +1,275 @@
+                    new ByteBuddy(ClassFileVersion.ofThisVm(ClassFileVersion.JAVA_V8))
+                            .subclass(Object.class)
+                            .modifiers(Modifier.FINAL | Modifier.PUBLIC)
+                            .name(className(interfaces))
+                            .defineConstructor(Visibility.PUBLIC)
+                            .withParameters(interfaceClass, InvocationEventHandler.class, InstrumentationFilter.class)
+                            .intercept(MethodCall.invoke(Object.class.getDeclaredConstructor())
+                                    .andThen(FieldAccessor.ofField("delegate").setsArgumentAt(0))
+                                    .andThen(FieldAccessor.ofField("invocationEventHandler").setsArgumentAt(1))
+                                    .andThen(FieldAccessor.ofField("instrumentationFilter").setsArgumentAt(2)))
+                            .implement(interfaces)
+                            .method(ElementMatchers.isToString())
+                            .intercept(MethodCall.invokeSelf().onField("delegate"));
+            List<Method> allMethods = new ArrayList<>();
+            for (Class<?> iface : interfaces) {
+                boolean allowDirectAccess = iface.isAssignableFrom(interfaceClass);
+                for (Method method : iface.getMethods()) {
+                    int index = allMethods.size();
+                    allMethods.add(method);
+                    // Retain tritium proxy detail where hashcode, equals, and toString cannot be instrumented.
+                    builder = builder.method(ElementMatchers.not(ElementMatchers.isHashCode()
+                            .or(ElementMatchers.isEquals())
+                            .or(ElementMatchers.isToString()))
+                            .and(ElementMatchers.is(method)))
+                            .intercept(Advice.withCustomMapping()
+                                    .bind(ByteBuddyInstrumentationAdvice.MethodIndex.class, index)
+                                    .to(ByteBuddyInstrumentationAdvice.class)
+                                    .wrap(allowDirectAccess
+                                            ? MethodCall.invokeSelf().onField("delegate").withAllArguments()
+                                            : MethodCall.invokeSelf()
+                                                    // Byte buddy doesn't seem to allow casting from fields, but
+                                                    // we can cast the result of a trivial call (in this case
+                                                    // Objects.requireNonNull) into the desired type.
+                                                    .onMethodCall(passThroughMethod().withField("delegate"))
+                                                    .withAllArguments()
+                                                    .withAssigner(Assigner.DEFAULT, Assigner.Typing.DYNAMIC)));
```

Comment:
Could simplify this a bit more and remove `boolean allowDirectAccess = iface.isAssignableFrom(interfaceClass);`, using the `Objects.requireNonNull` wrapper around all accesses of the delegate. I could go either way.